File: backend/app/routers/analyze.py
from fastapi import APIRouter, HTTPException
from app.models.schemas import AnalyzeRequest, AnalysisResult
from app.services.youtube_service import youtube_service
from app.services.sentiment_service import sentiment_service
from app.services.analytics_service import analytics_service
from app.services.insights_service import insights_service
import asyncio

router = APIRouter()


@router.post("/analyze", response_model=AnalysisResult)
async def analyze_video(request: AnalyzeRequest):
    video_id = request.extract_video_id()
    if not video_id:
        raise HTTPException(status_code=400, detail="Could not extract video ID from URL")

    try:
        metadata = await asyncio.to_thread(youtube_service.get_video_metadata, video_id)
    except (ValueError, RuntimeError) as e:
        raise HTTPException(status_code=404, detail=str(e))

    try:
        raw_comments = await asyncio.to_thread(youtube_service.fetch_comments, video_id)
    except RuntimeError as e:
        raise HTTPException(status_code=422, detail=str(e))

    if not raw_comments:
        raise HTTPException(status_code=404, detail="No comments found for this video")

    enriched = await asyncio.to_thread(sentiment_service.analyze_batch, raw_comments)

    sentiment_dist = analytics_service.compute_sentiment_distribution(enriched)
    emotion_dist = analytics_service.compute_emotion_distribution(enriched)
    engagement = analytics_service.compute_engagement(enriched)
    keywords = analytics_service.compute_keywords(enriched)
    top_positive = analytics_service.get_top_comments(enriched, "positive")
    top_negative = analytics_service.get_top_comments(enriched, "negative")
    sample = analytics_service.get_sample_comments(enriched)

    insights = insights_service.generate_insights(
        enriched, sentiment_dist.model_dump()
    )

    return AnalysisResult(
        video_id=video_id,
        video_title=metadata["title"],
        video_thumbnail=metadata["thumbnail"],
        total_comments_fetched=len(enriched),
        sentiment=sentiment_dist,
        emotions=emotion_dist,
        engagement=engagement,
        top_positive_comments=top_positive,
        top_negative_comments=top_negative,
        keywords=keywords,
        insights=insights,
        sample_comments=sample,
    )


@router.get("/health")
async def health():
    return {"status": "ok", "service": "YtAna API"}

File: backend/src/routes/analyze.js
const express = require('express');
const router = express.Router();
const youtubeService = require('../services/youtubeService');
const sentimentService = require('../services/sentimentService');
const analyticsService = require('../services/analyticsService');
const insightsService = require('../services/insightsService');

function extractVideoId(url) {
  const match = url.match(/(?:v=|youtu\.be\/|embed\/|shorts\/)([a-zA-Z0-9_-]{11})/);
  return match ? match[1] : null;
}

router.post('/analyze', async (req, res) => {
  const { url } = req.body;
  if (!url) return res.status(400).json({ detail: 'URL is required' });

  const videoId = extractVideoId(url);
  if (!videoId) return res.status(400).json({ detail: 'Could not extract video ID from URL' });

  let metadata;
  try {
    metadata = await youtubeService.getVideoMetadata(videoId);
  } catch (err) {
    return res.status(404).json({ detail: err.message });
  }

  let rawComments;
  try {
    rawComments = await youtubeService.fetchComments(videoId);
  } catch (err) {
    return res.status(422).json({ detail: err.message });
  }

  if (!rawComments.length) {
    return res.status(404).json({ detail: 'No comments found for this video' });
  }

  const enriched = sentimentService.analyzeBatch(rawComments);
  const sentiment = analyticsService.computeSentimentDistribution(enriched);
  const emotions = analyticsService.computeEmotionDistribution(enriched);
  const engagement = analyticsService.computeEngagement(enriched);
  const keywords = analyticsService.computeKeywords(enriched);
  const top_positive_comments = analyticsService.getTopComments(enriched, 'positive');
  const top_negative_comments = analyticsService.getTopComments(enriched, 'negative');
  const sample_comments = analyticsService.getSampleComments(enriched);
  const insights = insightsService.generateInsights(enriched, sentiment);

  res.json({
    video_id: videoId,
    video_title: metadata.title,
    video_thumbnail: metadata.thumbnail,
    total_comments_fetched: enriched.length,
    sentiment,
    emotions,
    engagement,
    top_positive_comments,
    top_negative_comments,
    keywords,
    insights,
    sample_comments,
  });
});

router.get('/health', (req, res) => {
  res.json({ status: 'ok', service: 'YtAna API' });
});

module.exports = router;
